### PR TITLE
Rust crypto: ensure we persist the key backup version

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes https://github.com/vector-im/element-web/issues/26259

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/crypto-api.ts [hunk 1]
```diff
@@ -374,9 +374,22 @@ export interface CryptoApi {
...
+    /**
+     * Store the backup decryption key.
+     *
+     * This should be called if the client has received the key from another device via secret sharing (gossiping).
+     *
+     * @param key - the backup decryption key
+     * @param version - the backup version corresponding to this decryption key
+     *
+     */
+    storeSessionBackupPrivateKey(key: Uint8Array, version?: string): Promise<void>;
```
Comment (richvdh) [inline src/crypto-api.ts:391]: shouldn't this be 

```suggestion
    storeSessionBackupPrivateKey(key: Uint8Array, version: string): Promise<void>;
```

?

Comment (BillCarsonFr) [reply src/crypto-api.ts:391]: Not sure, I though it was the way to do overloads

Comment (richvdh) [reply src/crypto-api.ts:391]: It's not required for overloads in an `interface`.

When you have overloads in a concrete class the the signature of the *implementation* must be compatible with all overloads (the "implementation signature"). But there is no implementation here.

https://www.typescriptlang.org/docs/handbook/2/functions.html#overload-signatures-and-the-implementation-signature

File: src/rust-crypto/rust-crypto.ts [hunk 2]
```diff
@@ -1005,12 +1005,20 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
      * @param key - the backup decryption key
+     * @param version - the backup version for this key. If not provided will fetch the current version from server.
      */
-    public async storeSessionBackupPrivateKey(key: Uint8Array): Promise<void> {
+    public async storeSessionBackupPrivateKey(key: Uint8Array, version?: string): Promise<void> {
         const base64Key = encodeBase64(key);
 
-        // TODO get version from backupManager
-        await this.olmMachine.saveBackupDecryptionKey(RustSdkCryptoJs.BackupDecryptionKey.fromBase64(base64Key), "");
+        const activeVersion = version ?? (await this.backupManager.requestKeyBackupVersion())?.version;
```
[the remark this answers is not in the document]
Comment (richvdh) [reply src/rust-crypto/rust-crypto.ts:1013]: We need to keep the old one working for old crypto. It doesn't mean we have to keep the old one working with rust crypto provided it works with the react-sdk.

File: src/crypto-api.ts [hunk 3]
```diff
@@ -372,10 +372,13 @@ export interface CryptoApi {
...
      *
      * This should be called if the client has received the key from another device via secret sharing (gossiping).
+     * It is the responsability of the caller to check that the decryption key is valid to the given backup version.
      *
      * @param key - the backup decryption key
+     * @param version - the backup version corresponding to this decryption key
+     *
      */
-    storeSessionBackupPrivateKey(key: Uint8Array): Promise<void>;
+    storeSessionBackupPrivateKey(key: Uint8Array, version: string): Promise<void>;
```
Comment (richvdh) [inline src/crypto-api.ts:381]: this is a breaking change. We need to continue to expose the old API, even if it only works on legacy crypto.

File: src/crypto/index.ts [hunk 4]
```diff
@@ -1302,7 +1302,8 @@ export class Crypto extends TypedEventEmitter<CryptoEvent, CryptoEventHandlerMap
         // make sure we have a Uint8Array, rather than a string
         if (typeof encodedKey === "string") {
             key = new Uint8Array(olmlib.decodeBase64(fixBackupKey(encodedKey) || encodedKey));
-            await this.storeSessionBackupPrivateKey(key);
+            // legacy crypto doesn't care about the version
+            await this.storeSessionBackupPrivateKey(key, "");
```
Comment (richvdh) [inline src/crypto/index.ts:1306]: yes. I don't think we need to change legacy crypto's behaviour at all here? It should continue to accept callls to `storeSessionBackupPrivateKey` with `version=undefined`.

File: src/rust-crypto/rust-crypto.ts [hunk 5]
```diff
@@ -1005,12 +1005,15 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
      * Implementation of {@link CryptoApi#storeSessionBackupPrivateKey}.
      *
      * @param key - the backup decryption key
+     * @param version - the backup version for this key. If not provided will fetch the current version from server.
```
Comment (richvdh) [inline src/rust-crypto/rust-crypto.ts:1008]: this seems wrong

File: src/crypto/CrossSigning.ts [hunk 6]
```diff
@@ -741,7 +741,7 @@ export async function requestKeysDuringVerification(
                 logger.info("Got key backup key, decoding...");
                 const decodedKey = decodeBase64(base64Key);
                 logger.info("Decoded backup key, storing...");
-                await client.crypto!.storeSessionBackupPrivateKey(Uint8Array.from(decodedKey));
+                await client.crypto!.storeSessionBackupPrivateKey(Uint8Array.from(decodedKey), "1");
```
Comment (richvdh) [inline src/crypto/CrossSigning.ts:744]: why "1" ???

File: src/client.ts [hunk 7]
```diff
@@ -3880,12 +3880,14 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
             }
             // Cache the key, if possible.
             // This is async.
-            this.cryptoBackend
-                .storeSessionBackupPrivateKey(privKey)
-                .catch((e) => {
-                    logger.warn("Error caching session backup key:", e);
-                })
-                .then(cacheCompleteCallback);
+            if (backupInfo.version != null) {
```
Comment (richvdh) [inline src/client.ts:3883]: I think we can move this check much earlier.  `version` is a required parameter on `GET /_matrix/client/v3/room_keys/keys/{roomId}` etc, so can throw an error at the top if it is not set.

**Pull request merged**